### PR TITLE
[WIP]CUE compiler to read CUE and encode into Go

### PR DESCRIPTION
**Pull request by poomethavornvanit**
Given CUE config, compile, and encode into GO structs based on bricks type

image

image

### CODE REVIEW

File: cue/client.cue [hunk 1]
```diff
@@ -0,0 +1,109 @@
+package bricks
+
+#service: {
```
Comment (spikelu2016) [inline cue/client.cue:3]: let's use Pascal case: service -> Service

File: cue/client.cue [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+    $bricks: type: "service"
```
Comment (spikelu2016) [inline cue/client.cue:4]: let's add a TODO comment to use hidden fields

File: cue/client.cue [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+	url: string
+	routes: [...#route]
+}
+
+#route: {
+    $bricks: type: "route"
+	name: string
+	description?: string
+	paths: [...string]
+    id: int
```
Comment (spikelu2016) [inline cue/client.cue:16]: why is id an int

File: main.go [hunk 4]
```diff
@@ -1,123 +1,60 @@
...
+	Selector string
+	Type     string
+}
 
-// 	return flow.RunnerFunc(func(t *flow.Task) error {
-// 		return nil
-// 	}), nil
-// }
+// Constants to map with CUE bricks.type selectors
+const ROUTE_TYPE = "route"
```
Comment (spikelu2016) [inline main.go:30]: nit: you can do something like
```
const (
  ROUTE_TYPE = "route"
  SERVICE_TYPE = "service"
 )
```

File: main.go [hunk 5]
```diff
@@ -1,123 +1,60 @@
...
-	// We should have a Greeter now! This feels like a normal interface
-	// implementation but is in fact over an RPC connection.
-	greeter := raw.(example.Greeter)
-	fmt.Println(greeter.Greet())
-
-	err = os.RemoveAll(pluginBuildPath)
-	if err != nil {
-		log.Fatal(err)
+	for _, m := range mappings {
+		if m.Type == ROUTE_TYPE {
```
Comment (spikelu2016) [inline main.go:46]: nit: we can use switch

File: main.go [hunk 6]
```diff
@@ -1,123 +1,60 @@
...
-// 	return flow.RunnerFunc(func(t *flow.Task) error {
-// 		return nil
-// 	}), nil
-// }
+// Constants to map with CUE bricks.type selectors
+const ROUTE_TYPE = "route"
+const SERVICE_TYPE = "service"
 
 func main() {
+	v, _ := compiler.Build(context.Background(), "./cue", nil)
```
Comment (spikelu2016) [inline main.go:34]: let's not suppress error. if there is an error, use Panic

File: main.go [hunk 7]
```diff
@@ -1,123 +1,60 @@
...
-	if err != nil {
-		log.Fatal(err)
-	}
+	var mappings []T
 
-	// Request the plugin
-	raw, err := rpcClient.Dispense("greeter")
-	if err != nil {
-		log.Fatal(err)
+	fields, _ := v.Fields()
```
Comment (spikelu2016) [inline main.go:38]: let's do panic on error

File: main.go [hunk 8]
```diff
@@ -1,123 +1,60 @@
...
-	// implementation but is in fact over an RPC connection.
-	greeter := raw.(example.Greeter)
-	fmt.Println(greeter.Greet())
-
-	err = os.RemoveAll(pluginBuildPath)
-	if err != nil {
-		log.Fatal(err)
+	for _, m := range mappings {
+		if m.Type == ROUTE_TYPE {
+			t := &Route{}
```
Comment (spikelu2016) [inline main.go:47]: do we expect users to write route without a service? if users write routes only in the context of a service, we only need to decode a service

File: main.go [hunk 9]
```diff
@@ -1,123 +1,60 @@
...
-		Logger:          logger,
-	})
-	defer client.Kill()
-
-	// Connect via RPC
-	rpcClient, err := client.Client()
-	if err != nil {
-		log.Fatal(err)
-	}
+	var mappings []T
```
Comment (spikelu2016) [inline main.go:36]: i would suggest we put all of the logic below in a function with input being compiler.Value and output being a list of services

Comment (spikelu2016) [reply main.go:36]: we don't really need mappings? i think we can do everything in a single for loop and skip cue instances without bricks.type field

File: cue/client.cue (hunk 3 above)
Comment (poomethavornvanit) [reply cue/client.cue:16]: got rid of id, was just for testing purposes

**Pull request merged**